packages/theme/src/cli/utilities/theme-environment/hot-reload/server.ts

```diff
@@ -0,0 +1,231 @@
+    return Promise.allSettled(
+      files.map(async ([fileKey, file]) => {
+        if (fileKey.endsWith('.json')) {
+          const content = file.value ?? (await ctx.localThemeFileSystem.read(fileKey))
+          if (content && typeof content === 'string') saveSectionsFromJson(fileKey, content)
+        }
+      }),
+    )
+  })
+}
+
+// --- SSE Hot Reload ---
+
+const eventEmitter = new EventEmitter()
+function emitHotReloadEvent(event: HotReloadEvent) {
+  eventEmitter.emit('hot-reload', event)
+}
+
+/**
+ * Adds endpoints to handle HotReload subscriptions and related events.
+ */
+export function getHotReloadHandler(theme: Theme, ctx: DevServerContext) {
+  return defineEventHandler((event) => {
+    const endpoint = event.path.split('?')[0]
+
+    if (endpoint === '/__hot-reload/subscribe') {
+      const eventStream = createEventStream(event)
+
+      eventEmitter.on('hot-reload', (event: HotReloadEvent) => {
+        eventStream.push(JSON.stringify(event)).catch((error: Error) => {
+          renderWarning({headline: 'Failed to send HotReload event.', body: error?.stack})
+        })
+      })
+
+      eventStream
+        .push(JSON.stringify({type: 'open', pid: String(process.pid)} satisfies HotReloadEvent))
+        .catch(() => {})
+
+      return eventStream.send().then(() => eventStream.flush())
+    } else if (endpoint === '/__hot-reload/render') {
+      const queryParams = getQuery(event)
+      const sectionId = queryParams['section-id']
+      const sectionKey = queryParams['section-template-name']
+
+      if (typeof sectionId !== 'string' || typeof sectionKey !== 'string') {
+        return
+      }
+
+      const sectionTemplate =
+        inMemoryTemplateFiles.has(sectionKey) && ctx.localThemeFileSystem.files.get(sectionKey)?.value
+
+      if (!sectionTemplate) {
+        renderWarning({headline: 'No template found for HotReload event.', body: `Template ${sectionKey} not found.`})
+        return
+      }
+
+      const replaceTemplates = {[sectionKey]: sectionTemplate}
+
+      // If a JSON file changed locally and updated the ID of a section,
+      // there's a chance the cloud won't know how to render a modified section ID.
+      // Therefore, we gather all the locally updated JSON files that reference
+      // the updated section ID and include them in replaceTemplates:
+      for (const fileKey of inMemoryTemplateFiles) {
+        if (fileKey.endsWith('.json')) {
+          for (const [_type, name] of sectionNamesByFile.get(fileKey) || []) {
+            // Section ID is something like `template_12345__<section-name>`:
+            if (sectionId.endsWith(`__${name}`)) {
+              const content = ctx.localThemeFileSystem.files.get(fileKey)?.value
+              if (content) replaceTemplates[fileKey] = content
+              continue
+            }
+          }
+        }
+      }
+
+      return render(ctx.session, {
+        path: '/',
+        query: [],
```

Comment:
I've fixed passing querystirngs but the issue seems to be something else. It even happens when there are no querystring at all, like in `/search` before searching anything.

Any other idea of why this could be happening? 🤔